images/kindnetd/cmd/kindnetd/subnets.go

```diff
@@ -0,0 +1,102 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"fmt"
+	"regexp"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/kubernetes"
+	"sigs.k8s.io/yaml"
+)
+
+type subnets struct {
+	pods     sets.String
+	services sets.String
+}
+
+func newSubnets() *subnets {
+	return &subnets{
+		pods:     sets.NewString(),
+		services: sets.NewString(),
+	}
+}
+
+func (s *subnets) All() []string {
+	return s.pods.Union(s.services).List()
+}
+
+func getNoMasqueradeSubnets(clientset *kubernetes.Clientset) []string {
+	s := newSubnets()
+
+	// try getting from kubeadm config
+	kubeadmSubnets := getKubeadmSubnets(clientset)
+	s.pods = s.pods.Union(kubeadmSubnets.pods)
+	s.services = s.services.Union(kubeadmSubnets.services)
+
+	// TODO: we have other fallback options for completeness including:
+	// - kube-apiserver flags
+	// - component config
+	// - defaults
+	return s.All()
+}
+
+const (
+	kubeadmServiceCIDRRegex = `(?m)^\s*serviceSubnet: (.*)\n`
+	kubeadmClusterCIDRRegex = `(?m)^\s*podSubnet: (.*)\n`
+)
+
+var (
+	kubeadmServiceCIDRRegexp = regexp.MustCompile(kubeadmServiceCIDRRegex)
+	kubeadmClusterCIDRRegexp = regexp.MustCompile(kubeadmClusterCIDRRegex)
+)
+
+func getKubeadmSubnets(clientset *kubernetes.Clientset) *subnets {
+	s := newSubnets()
+	configMap, err := clientset.CoreV1().ConfigMaps("kube-system").Get("kubeadm-config", metav1.GetOptions{})
+	if err != nil {
+		fmt.Println("error getting kubeadm config: " + err.Error())
+		return s
+	}
+	for _, data := range configMap.Data {
+		// ensure data is yaml formatted and does not contain comments
```

Comment:
dumb question, shouldn't this be guaranteed by client-go and kubeadm?

Reply:
or is for double-checking and for removing the comments?

Reply:
without the yamlify we should be able to work only with strings

Reply:
This is for completeness. What if the user edits the configmap?
Casting to byte slices is nearly free.

Reply:
good point, I imagine that other things will be affected but at least we are more or less safe

Reply:
(also it wasn't free before, but is in current go to the point the stdlib takes advantage of this to be simpler now, yamlify isn't free but the data is small and this makes us more robust to changes. It also ensures we know how the yaml is formatted, so we can match on a field being a line)